### PR TITLE
Last Mile Navigation: UnsubscribeWayPoints

### PR DESCRIPTION
**Pull request by dboltovskyi**
``` "GitHub issue link" ~ "https://github.com/SmartDeviceLink/sdl_core/issues/1741" ```

### CODE REVIEW

**Comment by dboltovskyi** [comment]
@GetmanetsIrina Please review

File: test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+local runner = require('user_modules/script_runner')
+local commonNavigation = require('test_scripts/API/Navigation/commonNavigation')
+
+--[[ Scenario ]]
+runner.Title("Preconditions")
+runner.Step("Clean environment", commonNavigation.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", commonNavigation.start)
+runner.Step("RAI, PTU", commonNavigation.registerAppWithPTU)
+runner.Step("Activate App", commonNavigation.activateApp)
+runner.Step("SubscribeWayPoints", commonNavigation.subscribeOnWayPointChange, { 1 })
```
Comment (AKalinich-Luxoft) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua]: @dboltovskyi what does this magic param {1} mean? Could it be replaced with named param of named constant?

File: test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+runner.Title("Preconditions")
+runner.Step("Clean environment", commonNavigation.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", commonNavigation.start)
+runner.Step("RAI, PTU", commonNavigation.registerAppWithPTU)
+runner.Step("Activate App", commonNavigation.activateApp)
+runner.Step("SubscribeWayPoints", commonNavigation.subscribeOnWayPointChange, { 1 })
+runner.Step("Is Subscribed", commonNavigation.isSubscribed)
+
+runner.Title("Test")
+runner.Step("IGNITION_OFF", commonNavigation.IGNITION_OFF)
```
Comment (AKalinich-Luxoft) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua]: @dboltovskyi why do you need ignition off here? Probably you could disable navi interface before first SDL start step

File: test_scripts/API/Navigation/UnsubscribeWayPoints/006_UnsubscribeWayPoints_GENERIC_ERROR_HMI_did_not_respond.lua [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+--[[ Local Functions ]]
+local function unsubscribeWayPoints(self)
+  local cid = self.mobileSession1:SendRPC("UnsubscribeWayPoints", {})
+  EXPECT_HMICALL("Navigation.UnsubscribeWayPoints")
+  :Do(function()
+      -- HMI does not respond
+    end)
+  self.mobileSession1:ExpectResponse(cid, { success = false, resultCode = "GENERIC_ERROR" })
+  commonTestCases:DelayedExp(11000)
```
Comment (AKalinich-Luxoft) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/006_UnsubscribeWayPoints_GENERIC_ERROR_HMI_did_not_respond.lua]: @dboltovskyi propose to replace with
`commonTestCases:DelayedExp(commonNavigation.timeout + 1000)`

File: test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+runner.Step("RAI, PTU", commonNavigation.registerAppWithPTU)
+runner.Step("Activate App", commonNavigation.activateApp)
+runner.Step("SubscribeWayPoints", commonNavigation.subscribeOnWayPointChange, { 1 })
+runner.Step("Is Subscribed", commonNavigation.isSubscribed)
+runner.Step("UnsubscribeWayPoints", commonNavigation.unsubscribeOnWayPointChange, { 1 })
+runner.Step("Is Unsubscribed", commonNavigation.isUnsubscribed)
+
+runner.Title("Test")
+runner.Step("UnsubscribeWayPoints for the same app IGNORED", unsubscribeWayPoints)
+runner.Step("Is Unsubscribed still", commonNavigation.isUnsubscribed)
```
Comment (AKalinich-Luxoft) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua]: @dboltovskyi propose to rename step to "Is still Unsubscribed" here and in other scripts

File: test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+--
+-- Description:
+-- In case:
+-- 1) mobile application sent valid UnsubscribeWayPoints_request to SDL
+-- 2) and this request is allowed by Policies
+-- 3) and there are other applications still subscribed on waypoints change npotifications
+--
+-- SDL must:
+-- 1) unsubscribe requesting application from waypoints change notifications
+-- 2) not send UnsubscribeWayPoints_reques to HMI
```
Comment (AKalinich-Luxoft) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua]: @dboltovskyi `UnsubscribeWayPoints_requesT`

File: test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua [hunk 6]
```diff
@@ -0,0 +1,39 @@
+---------------------------------------------------------------------------------------------------
+-- User story: https://github.com/smartdevicelink/sdl_requirements/issues/27
+-- Use case: https://github.com/smartdevicelink/sdl_requirements/blob/master/detailed_docs/embedded_navi/Unsubscribe_from_Destination_and_Waypoints.md
+-- Item: Use Case 1: Main Flow
+--
+-- Requirement summary:
+-- [UnsubscribeWayPoints] As a mobile app I want to be able to unsubscribes from getting notifications
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua]: @AKalinich-Luxoft Agreed

File: test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua (hunk 2 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua]: @AKalinich-Luxoft In order to perform Unsubscribe we have to be Subscribed.
So in order to perform subscription interface have to be available.

File: test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua (hunk 5 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua]: @AKalinich-Luxoft Agreed

File: test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua (hunk 4 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua]: @AKalinich-Luxoft Agreed

File: test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua (hunk 1 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua]: @AKalinich-Luxoft It was decided to add descriptions for all functions in common module

File: test_scripts/API/Navigation/UnsubscribeWayPoints/006_UnsubscribeWayPoints_GENERIC_ERROR_HMI_did_not_respond.lua (hunk 3 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/006_UnsubscribeWayPoints_GENERIC_ERROR_HMI_did_not_respond.lua]: @AKalinich-Luxoft Disagree
In this way total timeout will be 3s, but case requires 11s timeout

File: test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua (hunk 1 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua]: @AKalinich-Luxoft Please check 65b6521

File: test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua (hunk 4 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua]: @AKalinich-Luxoft Please check 65b6521

File: test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua (hunk 5 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/011_UnsubscribeWayPoints_2_apps.lua]: @AKalinich-Luxoft Please check 65b6521

File: test_scripts/API/Navigation/UnsubscribeWayPoints/002_UnsubscribeWayPoints_DISALLOWED_by_policy.lua [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+runner.Step("Clean environment", commonNavigation.preconditions)
+runner.Step("Start SDL, HMI, connect Mobile, start Session", commonNavigation.start)
+runner.Step("RAI, PTU", commonNavigation.registerAppWithPTU, { 1, ptuUpdateFunc })
+runner.Step("Activate App", commonNavigation.activateApp)
+runner.Step("SubscribeWayPoints", commonNavigation.subscribeOnWayPointChange, { 1 })
+runner.Step("Is Subscribed", commonNavigation.isSubscribed)
+
+runner.Title("Test")
+runner.Step("UnsubscribeWayPoints DISALLOWED by policy", unsubscribeWayPoints)
+
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/002_UnsubscribeWayPoints_DISALLOWED_by_policy.lua]: @GetmanetsIrina Agreed

File: test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+--
+-- Requirement summary:
+-- [UnsubscribeWayPoints] As a mobile app I want to be able to unsubscribes from getting notifications
+-- about any changes to the destination or waypoints.
+--
+-- Description:
+-- In case:
+-- 1) mobile application is already unsubscribed from WayPoints-related parameters
+-- 2) and the same mobile application sends UnsubscribeVehicleData_request to SDL
+--
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua]: @GetmanetsIrina Agreed

File: test_scripts/API/Navigation/UnsubscribeWayPoints/009_UnsubscribeWayPoints_unexpected_disconnect.lua [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+---------------------------------------------------------------------------------------------------
+--[[ Required Shared libraries ]]
+local runner = require('user_modules/script_runner')
+local commonNavigation = require('test_scripts/API/Navigation/commonNavigation')
+
+--[[ Local Functions ]]
+local function closeSession(self)
+  self.hmiConnection:SendNotification("BasicCommunication.OnAppUnregistered",
+    { unexpectedDisconnect = true, appID = commonNavigation.getHMIAppId() })
```
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/009_UnsubscribeWayPoints_unexpected_disconnect.lua]: @GetmanetsIrina Good catch

File: test_scripts/API/Navigation/UnsubscribeWayPoints/002_UnsubscribeWayPoints_DISALLOWED_by_policy.lua (hunk 7 above)
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/002_UnsubscribeWayPoints_DISALLOWED_by_policy.lua]: @GetmanetsIrina Please check 89d20e6

File: test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua (hunk 8 above)
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/003_UnsubscribeWayPoints_IGNORED_already_unsubscribed_app.lua]: @GetmanetsIrina Please check 89d20e6

File: test_scripts/API/Navigation/UnsubscribeWayPoints/009_UnsubscribeWayPoints_unexpected_disconnect.lua (hunk 9 above)
[the remark this answers is not in the document]
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/009_UnsubscribeWayPoints_unexpected_disconnect.lua]: @GetmanetsIrina Please check 89d20e6

File: test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua (hunk 1 above)
Comment (AByzhynar) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/001_UnsubscribeWayPoints_success_flow.lua]: @dboltovskyi As developr I support the idea to use named parameter instead of magic numbers.

File: test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua [hunk 10]
```diff
@@ -0,0 +1,56 @@
+---------------------------------------------------------------------------------------------------
```
Comment (AByzhynar) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua]: @dboltovskyi Please edit test case name. It contains error.

File: test_scripts/API/Navigation/UnsubscribeWayPoints/010_UnsubscribeWayPoints_Transfering_of_HMI_error_resultCode_to_mobile.lua [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+local failureResultCodes = {
+  "INVALID_DATA",
+  "GENERIC_ERROR",
+  "REJECTED",
+  "UNSUPPORTED_RESOURCE",
+  "IGNORED",
+  "DISALLOWED"
+}
+
+local unexpectedResultCodes = {
```
Comment (AByzhynar) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/010_UnsubscribeWayPoints_Transfering_of_HMI_error_resultCode_to_mobile.lua]: @dboltovskyi I propose not to hard code these codes but read needed from API.xml.
Any LUA parser can be used. This will be useful for all further code writers. 
[LuaXML](http://lua-users.org/wiki/LuaXml)

Comment (AByzhynar) [inline test_scripts/API/Navigation/UnsubscribeWayPoints/010_UnsubscribeWayPoints_Transfering_of_HMI_error_resultCode_to_mobile.lua]: @dboltovskyi I propose not to hard code these codes but read needed from API.xml. 
`user_modules/utils.lua` can be used

File: test_scripts/API/Navigation/commonNavigation.lua [hunk 12]
```diff
@@ -14,7 +14,7 @@ local commonSteps = require("user_modules/shared_testcases/commonSteps")
 local commonTestCases = require("user_modules/shared_testcases/commonTestCases")
 local SDL = require('SDL')
 
---[[ Local Variables ]]
```
Comment (AByzhynar) [inline test_scripts/API/Navigation/commonNavigation.lua]: @dboltovskyi Why did you remove `local`?

File: test_scripts/API/Navigation/UnsubscribeWayPoints/010_UnsubscribeWayPoints_Transfering_of_HMI_error_resultCode_to_mobile.lua (hunk 11 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/010_UnsubscribeWayPoints_Transfering_of_HMI_error_resultCode_to_mobile.lua]: @AByzhynar Accepted

File: test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua (hunk 10 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/UnsubscribeWayPoints/004_UnsubscribeWayPoints_UNSUPPORTED_RESOURCE_Navi_interface_is_not_avaliable.lua]: @AByzhynar Accepted

File: test_scripts/API/Navigation/commonNavigation.lua (hunk 12 above)
Comment (dboltovskyi) [reply test_scripts/API/Navigation/commonNavigation.lua]: @AByzhynar Because some of the variables are not local

**Comment by dboltovskyi** [comment]
@AByzhynar Please check 5209e5f
